### PR TITLE
Init ServiceRetriever Class

### PR DESCRIPTION
**Pull request by xpbowler**
# Pull Request Checklist

`ServiceRetriever` class for interacting with Anserini REST API. Currently only supports retrieval from pre-built indexes and k=10 (REST API does not accept k as a parameter currently).

Class variables:
- `host`
- dataset name (`dataset`)
- `retrieval_mode` (only DATASET is valid at the moment)
- `retrieval_method` (e.g., BM25)

`from_dataset_with_prebuilt_index()` takes in:
- `request`
- `dataset_name`
-  `retrieval_method`
- `k`  Anserini API currently does not support passing in k
-  `host` Defaults to `localhost::8081`

# PR Type

- [x] Feature

### CODE REVIEW

File: src/rank_llm/retrieve/service_retriever.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        Returns:
+            Request. Contains a query and list of candidates
+        Raises:
+            ValueError: If the retrieval mode is invalid or the result format is not as expected.
+        """
+        if self._retrieval_mode == RetrievalMode.DATASET:
+            query_url = parse.quote(request.query.text)
+            url = f"{self._host}/api/collection/{self._dataset}/search?query={query_url}&hits={str(k)}&qid={request.query.qid}"
+            response = requests.get(url)
+            if response.status_code == 200:
```
Comment (sahel-sh) [inline src/rank_llm/retrieve/service_retriever.py:103]: response.ok?

File: src/rank_llm/retrieve/service_retriever.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            raise ValueError(
+                f"Invalid retrieval method: {retrieval_method}. Please provide a specific retrieval method."
+            )
+        retriever = ServiceRetriever(
+            retrieval_mode=RetrievalMode.DATASET,
+            retrieval_method=retrieval_method,
+            dataset=dataset_name,
+            host=host,
+        )
+        return retriever.retrieve(k=k, request=request)
```
Comment (sahel-sh) [inline src/rank_llm/retrieve/service_retriever.py:83]: please return the retriever, the user can reuse the retriever for multiple retrieve requests rather than instantiating retriever every single time.  This way k and request args could be passed to retrieve function rather than init.

File: src/rank_llm/retrieve_and_rerank.py [hunk 3]
```diff
@@ -122,4 +145,4 @@ def retrieve_and_rerank(
                 for r in rerank_results
             ]
 
-    return rerank_results
+    return rerank_results[0]
```
Comment (sahel-sh) [inline src/rank_llm/retrieve_and_rerank.py:122]: why? this will break the batch mode.
If you want to return the first element in intractive mode, do it with the condition

File: src/rank_llm/retrieve_and_rerank.py [hunk 4]
```diff
@@ -48,9 +54,13 @@ def retrieve_and_rerank(
             keys=openai_keys,
             **(get_azure_openai_args() if use_azure_openai else {}),
         )
-    elif "vicuna" in model_path.lower() or "zephyr" in model_path.lower():
+    elif model_path.lower()=="rank_zephyr" or model_path.lower()=="rank_vicuna":
```
Comment (sahel-sh) [inline src/rank_llm/retrieve_and_rerank.py:58]: why this change? It will break the code if we have local checkpoints of finetuned models. 
Within the existing elif, you can handle rank_zephyr and rank_vicuna as special cases.

File: src/rank_llm/retrieve/service_retriever.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+
+class ServiceRetriever:
+    def __init__(
+        self,
+        dataset: Union[str, List[str], List[Dict[str, Any]]],
+        host: str,
+        retrieval_mode: RetrievalMode = RetrievalMode.DATASET,
+        retrieval_method: RetrievalMethod = RetrievalMethod.BM25,
+        # index_path: str = None,
```
Comment (sahel-sh) [inline src/rank_llm/retrieve/service_retriever.py:28]: please clean up commented code

File: src/rank_llm/retrieve/service_retriever.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        return self.value
+
+
+class ServiceRetriever:
+    def __init__(
+        self,
+        dataset: Union[str, List[str], List[Dict[str, Any]]],
+        host: str,
+        retrieval_mode: RetrievalMode = RetrievalMode.DATASET,
+        retrieval_method: RetrievalMethod = RetrievalMethod.BM25,
```
Comment (sahel-sh) [inline src/rank_llm/retrieve/service_retriever.py:27]: This should be part of the url params for the anserini get request similar to dataset.
If anserini rest api does not expose retrievers other than bm25, please leave a todo and raise an error similar to retrieval method

**Review by sahel-sh — CHANGES_REQUESTED** [review]

File: src/rank_llm/retrieve/service_retriever.py (hunk 6 above)
Comment (xpbowler) [reply src/rank_llm/retrieve/service_retriever.py:27]: If the class has no member variables, then should we still need a ServiceRetriever class?

File: src/rank_llm/rerank/rankllm.py [hunk 7]
```diff
@@ -331,6 +334,8 @@ def covert_doc_to_prompt_content(self, doc: Dict[str, Any], max_length: int) ->
             content = doc["segment"]
         elif "contents" in doc:
             content = doc["contents"]
+        elif "body" in doc:
+            content = doc["body"]
```
Comment (ronakice) [inline src/rank_llm/rerank/rankllm.py:338]: Why did you reintroduce this?

File: src/rank_llm/retrieve_and_rerank.py [hunk 8]
```diff
@@ -64,62 +83,55 @@ def retrieve_and_rerank(
...
     reranker = Reranker(agent)
     for pass_ct in range(num_passes):
         print(f"Pass {pass_ct + 1} of {num_passes}:")
-        rerank_results = reranker.rerank_batach(
+        rerank_results = reranker.rerank_batch(
             requests,
-            rank_end=top_k_candidates,
-            window_size=min(window_size, top_k_candidates),
+            rank_end=top_k_retrieve,
+            window_size=min(window_size, top_k_rerank),
```
Comment (ronakice) [inline src/rank_llm/retrieve_and_rerank.py:123]: should be `min(window_size, top_k_retrieve)`

File: src/rank_llm/retrieve/service_retriever.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        
+        if not retrieval_method:
+            raise "Please provide a retrieval method."
+        
+        if retrieval_method == RetrievalMethod.UNSPECIFIED:
+            raise ValueError(f"Invalid retrieval method: {retrieval_method}. Please provide a specific retrieval method.")
+        
+    def retrieve(
+        self, 
+        dataset: Union[str, List[str], List[Dict[str, Any]]],
```
Comment (sahel-sh) [inline src/rank_llm/retrieve/service_retriever.py:47]: please update the type hint to be str only for now

File: src/rank_llm/retrieve/service_retriever.py (hunk 6 above)
Comment (sahel-sh) [reply src/rank_llm/retrieve/service_retriever.py:18]: For now since `retrieval_method` and `retrieval_mode` have fixed values, the `retrieve` function doesn't need to know about them but eventually they need to become member variables for `retrieve` to access them.

File: src/rank_llm/retrieve_and_rerank.py [hunk 10]
```diff
@@ -36,9 +39,16 @@ def retrieve_and_rerank(
     index_path: str = None,
     topics_path: str = None,
     index_type: str = None,
+    interactive: bool = False,
+    host: str = "http://localhost:8081",
+    exec_summary: bool = False,
```
Comment (sahel-sh) [inline src/rank_llm/retrieve_and_rerank.py:44]: please rename `exec_summary` to `populate_exec_summary` here and everywhere else to better align with the intended boolean value of the arg.

File: src/rank_llm/rerank/rank_gpt.py [hunk 11]
```diff
@@ -127,7 +127,7 @@ def run_llm(
         prompt: Union[str, List[Dict[str, str]]],
         current_window_size: Optional[int] = None,
     ) -> Tuple[str, int]:
-        model_key = "engine" if self.use_azure_ai else "model"
+        model_key = "model"
```
Comment (sahel-sh) [inline src/rank_llm/rerank/rank_gpt.py:130]: why this change?

Comment (xpbowler) [reply src/rank_llm/rerank/rank_gpt.py:130]: This was Ronak

File: src/rank_llm/retrieve/service_retriever.py (hunk 6 above)
Comment (xpbowler) [reply src/rank_llm/retrieve/service_retriever.py:18]: Sure

File: src/rank_llm/rerank/rankllm.py (hunk 7 above)
Comment (xpbowler) [reply src/rank_llm/rerank/rankllm.py:338]: because the doc object in Anserini's API response only has a body section rather than text,segment, or contents

File: src/rank_llm/rerank/rank_gpt.py (hunk 11 above)
Comment (sahel-sh) [reply src/rank_llm/rerank/rank_gpt.py:130]: @ronakice PTAL

File: src/rank_llm/retrieve_and_rerank.py [hunk 12]
```diff
@@ -64,62 +83,55 @@ def retrieve_and_rerank(
...
-                print(f"Skipping evaluation as {dataset} is not in TOPICS.")
         if num_passes > 1:
             requests = [
-                Request(copy.deepycopy(r.query), copy.deepcopy(r.candidates))
+                Request(copy.deepcopy(r.query), copy.deepcopy(r.candidates))
                 for r in rerank_results
             ]
-
+    print(f"Reranking with {num_passes} passes complete!")
+    rerank_results[0].candidates = rerank_results[0].candidates[:top_k_rerank]
```
Comment (sahel-sh) [inline src/rank_llm/retrieve_and_rerank.py:136]: Please truncate the rerank results for all indexes rather than the first one only

File: src/rank_llm/api/server.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        raise ValueError(f"Unsupported model: {model}")
+
+    @app.route('/api/model/<string:model_path>/index/<string:dataset>/<string:host>', methods=['GET'])
+    def search(model_path, dataset, host):
+
+        query = request.args.get('query',type=str)
+        top_k_retrieve = request.args.get('hits_retriever',default=20,type=int)
+        top_k_rerank = request.args.get('hits_reranker',default=5,type=int)
+        qid = request.args.get('qid',default=None,type=str)
+        num_passes = request.args.get('num_passes',default=1,type=int)
```
Comment (sahel-sh) [inline src/rank_llm/api/server.py:73]: These lines seem to need formatting, please install and run black for formatting.

File: src/rank_llm/retrieve/service_retriever.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        Returns:
+            Request. Contains a query and list of candidates
+        Raises:
+            ValueError: If the retrieval mode is invalid or the result format is not as expected.
+        """
+
+        parsed_query = parse.quote(request.query.text)
+        url = f"{host}/api/index/{dataset}/search?query={parsed_query}&hits={str(k)}&qid={request.query.qid}"
+
+        response = requests.get(url)
```
Comment (sahel-sh) [inline src/rank_llm/retrieve/service_retriever.py:70]: Sorry, I missed this earlier, when anserini get request fails, we should not raise a value error. Please replace the code block with something like this:
Ideally we want to handle different error cases differently, but for now a catch all and raise is ok.
```
try:
    r = requests.get(url,timeout=3)
    r.raise_for_status()
except Exception as e:
    raise type(e)('Failed to retrieve data from Anserini server: ' + e.message)

**Review by sahel-sh — APPROVED** [review]

**Comment by xpbowler** [comment]
@sahel-sh linted + added error handling

File: src/rank_llm/retrieve_and_rerank.py [hunk 15]
```diff
@@ -64,62 +83,62 @@ def retrieve_and_rerank(
...
-                    ["-c", "-m", "ndcg_cut.1", TOPICS[dataset], file_name]
-                )
-                EvalFunction.eval(
-                    ["-c", "-m", "ndcg_cut.5", TOPICS[dataset], file_name]
-                )
-                EvalFunction.eval(
-                    ["-c", "-m", "ndcg_cut.10", TOPICS[dataset], file_name]
-                )
-            else:
-                print(f"Skipping evaluation as {dataset} is not in TOPICS.")
```
Comment (ronakice) [inline src/rank_llm/retrieve_and_rerank.py:118]: @xpbowler add this back?

File: src/rank_llm/rerank/rank_gpt.py (hunk 11 above)
Comment (ronakice) [reply src/rank_llm/rerank/rank_gpt.py:130]: I think the engine is now just model? I was getting errors when I try to run it in the newest version.

**Review by ronakice — APPROVED** [review]
